tempesta_fw/client.c

```diff
@@ -28,119 +28,158 @@
+ * and @user_agent.
  *
  * The returned TfwClient reference must be released via tfw_client_put()
  * when the @sk is closed.
- *
- * TODO #488 (#100): evict connections and/or clients and drop their accounting.
- * For now a client is freed immediately when the last of its connections is
- * closed, probably we should evict clients after some timeout to keep their
- * classifier statistic for following sessions...
+ * TODO #515 employ eviction strategy for the table.
  */
 TfwClient *
-tfw_client_obtain(struct sock *sk, void (*init)(TfwClient *))
+tfw_client_obtain(TfwAddr addr, TfwAddr *cli_addr, TfwStr *user_agent,
+		  void (*init)(TfwClient *))
 {
+	TfwClientEntry *ent;
 	TfwClient *cli;
 	CliHashBucket *hb;
-	struct hlist_node *tmp;
 	unsigned long key;
-	TfwAddr addr;
+	size_t len;
+	TdbRec *rec;
+	bool is_new;
+	int conn_users;
+	time_t curr_time = tfw_current_timestamp();
 
-	ss_getpeername(sk, &addr);
 	key = hash_calc((const char *)&addr.sin6_addr,
 			sizeof(addr.sin6_addr));
 
 	hb = &cli_hash[hash_min(key, CLI_HASH_BITS)];
 
 	spin_lock(&hb->lock);
 
-	hlist_for_each_entry_safe(cli, tmp, &hb->list, hentry)
-		if (ipv6_addr_equal(&addr.sin6_addr,
-		                    &cli->addr.sin6_addr))
-			goto found;
-
-	if (!(cli = kmem_cache_alloc(cli_cache, GFP_ATOMIC | __GFP_ZERO))) {
-		spin_unlock(&hb->lock);
+	if (cli_addr)
+		key ^= hash_calc((const char *)&cli_addr->sin6_addr,
+				 sizeof(cli_addr->sin6_addr));
+	if (user_agent)
+		key ^= tfw_hash_str_len(user_agent, 256);
+
+	len = sizeof(*ent);
+	rec = tdb_rec_get_alloc(client_db, key, &len, &tfw_client_addr_eq,
+				&is_new, &addr.sin6_addr);
+	if (!rec) {
+		TFW_WARN("cannot allocate TDB space for client\n");
 		return NULL;
 	}
+	ent = (TfwClientEntry *)rec->data;
+	cli = &ent->cli;
+
+	if (is_new) {
+		ent->hb_lock = &hb->lock;
+		ent->expires = LONG_MAX;
+		if (init)
+			init(cli);
+
+		atomic_set(&cli->conn_users, 1);
+		TFW_INC_STAT_BH(clnt.online);
+
+		tfw_peer_init((TfwPeer *)cli, &addr);
+
+		TFW_DBG("new client: cli=%p\n", cli);
+		TFW_DBG_ADDR("client address", &cli->addr, TFW_NO_PORT);
+	} else {
+		if (curr_time > ent->expires) {
+			bzero_fast(&cli->class_prvt, sizeof(cli->class_prvt));
+			if (init)
+				init(cli);
+		}
 
-	tfw_peer_init((TfwPeer *)cli, &addr);
-	hlist_add_head(&cli->hentry, &hb->list);
-	cli->hb_lock = &hb->lock;
-	if (init)
-		init(cli);
+		ent->expires = LONG_MAX;
 
-	atomic64_inc(&act_cli_n);
-	TFW_INC_STAT_BH(clnt.online);
-	TFW_DBG("new client: cli=%p\n", cli);
-	TFW_DBG_ADDR("client address", &cli->addr, TFW_NO_PORT);
+		conn_users = atomic_inc_return(&cli->conn_users);
+		if (conn_users == 1)
+			TFW_INC_STAT_BH(clnt.online);
 
-found:
-	atomic_inc(&cli->conn_users);
+		TFW_DBG("client was found in tdb\n");
+	}
 
-	TFW_DBG2("client %p, conn_users=%d\n",
-		 cli, atomic_read(&cli->conn_users));
+	TFW_DBG2("client %p, conn_users=%d\n", cli, conn_users);
+
+	if (!is_new)
+		tdb_rec_put(rec);
 
 	spin_unlock(&hb->lock);
```

Comment:
Why do we put the record? If there is 2 connections in the system referring to the same client, then they both must keep read locks for the TDB bucket containing the client record.

What's the paired lock acquisition for `spin_unlock(&hb->lock);`?

Reply:
The record doesn't change its location in TDB, since it is more than TDB_HTRIE_MINDREC, and we need to unlock the bucket with the client as soon as possible.